Refactored possibly dead code to allow const to be added to function parameter

Waffl3x opened the pull request:
Please see the commit message on 38012bcd20c343a12c324f85326f51c5baefe002 as I can't be certain this doesn't break anything.

I would have preferred to simply remove the block that I suspect is dead code, if that is an option, then my changes to VkCreateInstance() are unnecessary. It more or less seems to me that override_layer_present might have a bigger role in loader_scan_for_implicit_layers(), but tests still pass even when it's edited to be set to false there too, so I'm not exactly sure what to think.

ci-tester-lunarg commented:
Author Waffl3x not on autobuild list. Waiting for curator authorization before starting CI build.

charles-lunarg commented:
loader.c:961 is where `override_layer_present` is used and it definitely isn't dead code.
While I do appreciate the intent (making parameters const) this PR unfortunately is not useful due to it potentially making breaking changes.
Set a break point in `OverrideMetaLayer.OlderMetaLayerWithNewerInstanceVersion` at loader.c:961 and you'll see exactly where it is necessary.

Waffl3x commented:
I found the test and I will be sure to set a breakpoint there when I continue to investigate this. I don't understand why the test passed despite the code block mentioned being completely commented out, but perhaps I just did the old forgetting to save before compiling.

Just in case you misunderstood me, I wasn't trying to say that testing for override layers is dead code, but merely that specific block that compares against it.
In addition, the last thing I want to do is propose breaking changes, which is why I made sure that I added a new block to set override_layer_present even though as far as I can tell the flag isn't /actually/ being used. I suspect that the flag was made redundant through other changes and it wasn't noticed that it no longer gets used.

I will make sure to investigate this fully before I press the issue more though. I hadn't thought of using a debugger while running tests, using it I think I'll be able to demonstrate that either the code has since been made redundant, or at lesat that my changes don't cause any breakage. If I find that my changes or removing the code does cause breakage I'll figure out why the tests don't fail and come back with that.

charles-lunarg commented:
Just keep me posted, it'd be nice if there is some real dead code or at least some 'this shouldn't be like this' code.

And as far as debugging goes, the test always enable VK_LOADER_DEBUG=all, which is very helpful when reading through the log of 'what the loader has done'.

Waffl3x commented:
I'm having trouble getting breakpoints to work, I'm running ctest through gdb, is there anything special I'm supposed to be doing or is it expected that I can just do "break file:line" and it should work automatically?

As a side note, I read the tests to make sure it actually passes through what I'm breaking on and I discovered that this particular test does not test any code that I changed. I skimmed the rest of the OverrideMetaLayer tests and none of them call vkCreateInstance, and the function I changed loader_scan_for_layers only gets called (with a non null inst parameter) from vkCreateInstance.

Once I understand more of the big picture here I will look into writing a test for that, but I believe I need to get breakpoints working while running tests before I bother.

charles-lunarg commented:
Try running the ‘test_regression’ executable directly, ctest is spawning processes which don’t know about the debugger .

Waffl3x commented:
Alright, all the suffering and learning new things was worthwhile, I found that during test `OverrideMetaLayer.OlderMetaLayerWithNewerInstanceVersion` at `loader.c:961` the value of `enable` is always true if `inst` is non null, so at least in the context of this test, `inst->override_layer_present`'s value does not have an impact on control flow.

I will look deeper into it to see if I can fabricate any cases where the value is necessary, but I currently suspect there wont be because of how the rest of the function looks to me. I'll get started on that tomorrow.

Thanks for being so patient with me, I know I'm a little out of my depth but I like to jump straight into the deep end.

charles-lunarg commented:
To document some discussion had elsewhere, the place where `override_layer_present` is read from is necessary because of implicit layers which have an enable environment variable. It would be prudent to add a test layer which makes use of that enable environment variable to the tests in question.

But more importantly, I'd love to see this PR be turned into a 'remove override_layer_present  from loader_instance and make the variable be passed in explicitly'. It doesn't need to live inside loader_instance since it's only used in one place and not outside of vkCreateInstance at all

The test code to add to OverrideMetaLayer.OlderMetaLayerWithNewerInstanceVersion and similar.
```cpp
    const char* enable_implicit_layer_name = "VK_LAYER_EnableTestLayer";
    env.add_implicit_layer(ManifestLayer{}
                               .set_file_format_version(ManifestVersion(1, 2, 0))
                               .add_layer(ManifestLayer::LayerDescription{}
                                              .set_name(enable_implicit_layer_name)
                                              .set_enable_environment("GottaEnableMe")
                                              .set_disable_environment("DontDisableMe")
                                              .set_lib_path(TEST_LAYER_PATH_EXPORT_VERSION_2)
                                              .set_api_version(VK_MAKE_API_VERSION(0, 1, 1, 0))),
                           "enable_implicit_test_layer.json");
```

ci-tester-lunarg commented:
Author Waffl3x not on autobuild list. Waiting for curator authorization before starting CI build.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 53748.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1689 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1689 passed.

charles-lunarg commented:
Due to a huge amount of code change since this PR was created, I'm going to close it. (specifically the loader settings PR which modifies the layer lists logic substantially.)

Waffl3x commented:
Just for the record, I hadn't intended the commit to my repo to also add to this pull request, so I have no objections to closing this (and in hindsight I should have done it myself.)
